File: src/main/swift/g0001_0100/s0010_regular_expression_matching/Solution.swift
// #Hard #Top_Interview_Questions #String #Dynamic_Programming #Recursion #Udemy_Dynamic_Programming
// #Big_O_Time_O(m*n)_Space_O(m*n) #2024_06_18_Time_3_ms_(91.89%)_Space_16.2_MB_(58.11%)

class Solution {
    func isMatch(_ string: String, _ pattern: String) -> Bool {
        let stringArray = Array(string)
        let patternArray = Array(pattern)

        let stringLength = stringArray.count
        let patternLength = patternArray.count

        var table = Array(
            repeating: Array(repeating: false, count: patternLength + 1),
            count: stringLength + 1
        )

        table[stringLength][patternLength] = true

        for s in (0...stringLength).reversed() {
            for p in (0..<patternLength).reversed() {
                let firstMatch = s < stringLength 
                    && (stringArray[s] == patternArray[p] || patternArray[p] == ".")

                if p + 1 < patternLength && patternArray[p + 1] == "*" {
                    table[s][p] = firstMatch && table[s + 1][p] || table[s][p + 2]
                } else {
                    table[s][p] = firstMatch && table[s + 1][p + 1]
                }
            }
        }

        return table[0][0]
    }
}

File: src/main/rust/g0001_0100/s0010_regular_expression_matching/Solution.rs
// #Hard #Top_Interview_Questions #String #Dynamic_Programming #Recursion #Udemy_Dynamic_Programming
// #Big_O_Time_O(m*n)_Space_O(m*n) #2024_09_04_Time_0_ms_(100.00%)_Space_2.1_MB_(25.55%)

impl Solution {
    pub fn is_match(s: String, p: String) -> bool {
        is_match_dp(s.as_bytes(), p.as_bytes())
    }
}

pub fn is_match_dp(s: &[u8], p: &[u8]) -> bool {
    let mut grid = vec![vec![false; s.len() + 1]; p.len() + 1];
    grid[p.len()][s.len()] = true;

    let mut pi = p.len();
    while pi > 0 {
        pi -= 1;

        let mstar = pi > 0 && p[pi] == b'*';
        if mstar {
            pi -= 1;
        }

        for si in 0..=s.len() {
            let is_match = if mstar {
                let mut si2 = si;
                loop {
                    if grid[pi + 2][si2] {
                        break true;
                    }
                    if si2 < s.len() && (p[pi] == b'.' || s[si2] == p[pi]) {
                        si2 += 1;
                        continue;
                    }
                    break false;
                }
            } else {
                si < s.len() && (p[pi] == b'.' || s[si] == p[pi]) && grid[pi + 1][si + 1]
            };
            grid[pi][si] = is_match;
        }
    }

    grid[0][0]
}
